Add backend support for starring messages

yourFriendWes opened the pull request:
### Why: 
We would like chat users to be able to star messages and refer back to a history of starred messages.

### What: 
This PR lays the backend groundwork for associating a user with starred messages. 

It creates a join table between Users and Messages called SharedMessages, and adds a schema for SharedMessages with `create` and `delete functions`.

### Verification Steps:
See https://github.com/oberon-chat/chat-client/pull/4 for combined verification steps.

yourFriendWes committed:
Create migration for starred messages table

yourFriendWes committed:
Create migration for starred messages table

yourFriendWes committed:
Create Schema for StarredMessages

yourFriendWes committed:
Create Schema for StarredMessages

yourFriendWes committed:
Add many_to_many relationships for starred messages

yourFriendWes committed:
Add many_to_many relationships for starred messages

yourFriendWes committed:
Fix many_to_many declarations

yourFriendWes committed:
Fix many_to_many declarations

yourFriendWes committed:
Skip pending tests

yourFriendWes committed:
Skip pending tests

yourFriendWes committed:
WIP commit - test skeleton for starred messages

yourFriendWes committed:
WIP commit - test skeleton for starred messages

yourFriendWes committed:
Remove test skeleton

- unit tests were not meaningful

yourFriendWes committed:
Remove test skeleton

- unit tests were not meaningful

yourFriendWes committed:
clean up whitespace

yourFriendWes committed:
clean up whitespace

chrislaskey commented on apps/chat_server/lib/chat_server/schemas/message.ex:
```diff
@@ -10,6 +10,8 @@ defmodule ChatServer.Schema.Message do
     belongs_to :room, ChatServer.Schema.Room
     belongs_to :user, ChatServer.Schema.User
 
+    many_to_many :users, ChatServer.Schema.User, join_through: "starred_messages"
```
Is there a more descriptive name we could use? I wouldn't necessarily make the mental connection between `%Message{users: ...}` pointing to "Users who have starred this message".

What if we used something like `stargazers`? Or `starred_users` or `star_users`? Spitballing here, additional ideas welcome!

chrislaskey commented on apps/chat_server/lib/chat_server/schemas/user.ex:
```diff
@@ -11,8 +11,9 @@ defmodule ChatServer.Schema.User do
-
     timestamps()
+
+    many_to_many :messages, ChatServer.Schema.Message, join_through: "starred_messages"
```
Same as above, `messages` feels like a too generic name. Maybe `starred_messages`?

yourFriendWes replied:
ah okay, I didn't make the connection that I have the freedom to have the name differ from the schema here. `starred_messages` sounds great. The name is a little harder to figure out on the Message side, maybe `users_following`?

chrislaskey commented:
Looks good so far, I'd add some schema tests to confirm the behavior.

For example, I think the `starred_messages` table could end up with orphaned records when a message is deleted. To fix it a `on_delete` option may need to be set on the `many_to_many` definition.

Best way to experiment and prove it out is with a new `starred_messages` test file. There should be some examples of basic record creating tests in the `groups` schema on the latest `master` branch :)

yourFriendWes committed:
rename users relationship

yourFriendWes committed:
rename users relationship

yourFriendWes committed:
Update starred message associations to have delete

yourFriendWes committed:
Update starred message associations to have delete

yourFriendWes committed:
WIP add test for starred message

- WIP because new test failures seen in user_test now.

yourFriendWes committed:
WIP add test for starred message

- WIP because new test failures seen in user_test now.

chrislaskey committed:
Fix tests

chrislaskey committed:
Fix tests

yourFriendWes committed:
Merge pull request #1 from chrislaskey/add-starred-messages

Fix tests

yourFriendWes committed:
WIP commit - Add Starred Message Contexts

- Does not compile, pushing to get team’s thoughts

yourFriendWes committed:
WIP commit Attempt at preloading starred_messages with users

- not compiling yet, just pushing for team review and thoughts

yourFriendWes committed:
Merge branch 'add-starred-messages' of github.com:yourFriendWes/chat-server into add-starred-messages

yourFriendWes committed:
Remove non-functional preload starred_messages

yourFriendWes committed:
Fix starred_message queries

yourFriendWes committed:
Rewrite starred message contexts

yourFriendWes committed:
WIP commit: Add handler for starred_message:create

-

yourFriendWes committed:
WIP commit

yourFriendWes committed:
Fix starred_messages_query

yourFriendWes committed:
Move get_starred_message socket push

yourFriendWes committed:
Remove unnecessary starred_message logic

yourFriendWes committed:
Add starred_message:create in rooms channel

yourFriendWes committed:
Remove unneccesary aliases

yourFriendWes committed:
Add handle_in for starred_message:delete

yourFriendWes committed:
WIP commit: nonfunctional context functions for star deletion

yourFriendWes committed:
Fix errors in casting associations and star_message_query

yourFriendWes committed:
Format data sent to StarMessage.create to be map

- w

yourFriendWes committed:
add return statements to starred_message delete and create handlers

yourFriendWes committed:
Change method calls that load StarredMessage

yourFriendWes committed:
Add Migration to include id in StarredMessage

yourFriendWes committed:
Fix syntax of delete success message to return map

chrislaskey committed:
Merge remote-tracking branch 'upstream/master' into add-starred-messages

yourFriendWes committed:
Merge pull request #2 from chrislaskey/add-starred-messages

Merge upstream/master into Starred Messages Branch

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/create_starred_message.ex:
```diff
@@ -0,0 +1,28 @@
+    else
+      error ->
+        Logger.debug "Error starring message " <> inspect(error)
+          {:error, "Error starring message"}
```
Should have the same indentation as `Logger` line.

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/create_starred_message.ex:
```diff
@@ -0,0 +1,28 @@
+  def call(message_id, user) do
+    message = Schema.Message.get(message_id)
+    with {:ok, record} <- create_record(message, user) do
+        {:ok, Schema.StarredMessage.preload(record, [:message, :user])}
```
Should be indented 2 spaces deeper than `with`. Looks like it's currently 4.

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/create_starred_message.ex:
```diff
@@ -0,0 +1,28 @@
+
+  def call(message_id, user) do
+    message = Schema.Message.get(message_id)
+    with {:ok, record} <- create_record(message, user) do
```
Line 7 can be moved into the `with` statement:

```
with message <- Schema.Message.get(message_id),
     {:ok, record} <- create_record(message, user) do
```

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/delete_starred_message.ex:
```diff
@@ -0,0 +1,19 @@
+    star_id = Map.get(starred_message, :id)
+
+    with {:ok, _record} <- Schema.StarredMessage.delete(starred_message) do
+        {:ok, %{id: star_id}}
```
Same indentations here, looks like 4 but should be 2 deeper than `with` on line 11.

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/delete_starred_message.ex:
```diff
@@ -0,0 +1,19 @@
+    else
+      error ->
+        Logger.debug "Error deleting starred message" <> inspect(error)
+          {:error, "Error deleting starred message"}
```
Same here too, the tuple on line 16 should be indented the same as `Logger` on line 15.

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/delete_starred_message.ex:
```diff
@@ -0,0 +1,19 @@
+    starred_message = Schema.StarredMessage.find_by_message_and_user(message, user)
+    star_id = Map.get(starred_message, :id)
+
+    with {:ok, _record} <- Schema.StarredMessage.delete(starred_message) do
```
If only one clause is being evaluated, a `case` statement is more clear and concise.

The value of a `with` statement is it evaluates multiple clauses and guarantees they have returned the expected values.

In this instance, the function is making multiple calls. It could be rewritten where each is a `with` clause:

```elixir
def call(message_id, user) do
  with message <- Schema.Message.get(message_id),
       starred_message <- Schema.StarredMessage.find_by_message_and_user(message, user),
       {:ok, _record} <- Schema.StarredMessage.delete(starred_message) do
    {:ok, %{id: starred_message.id}}
  else
    # ...
  end
end
```

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/get_starred_messages.ex:
```diff
@@ -0,0 +1,11 @@
+  defp get_starred_messages(user) do
+    Schema.StarredMessage.get_by_user(user)
+  end
+end
```
This context looks good. Two thoughts to take it further:

1. I've been refactoring context objects to always return a tuple, e.g. either `{:ok, result}` or `{:error, msg}`. This makes it a lot easier to chain context objects in `case` and `with` statements and ensure a valid result is being returned without introducing lots of `nil` checks downstream. For an example of that, see: [GetSubscription](https://github.com/oberon-chat/chat-server/blob/master/apps/chat_server/lib/chat_server/contexts/subscriptions/get_subscription.ex).

2. A context object could be called from multiple places, e.g. Websocket and GraphQL. The returned `%{starred_messages: data}` map is a Websocket implementation detail, and should be handled in the channel file instead of in the context. The context should return just the data, or ideally, `{:ok, data}` | `{:error, msg}`

chrislaskey commented on apps/chat_server/lib/chat_server/schemas/message.ex:
```diff
@@ -10,6 +10,8 @@ defmodule ChatServer.Schema.Message do
     belongs_to :room, Schema.Room
     belongs_to :user, Schema.User
 
+    many_to_many :users_following, Schema.User, join_through: "starred_messages, on_delete: :delete_all, on_replace: :delete"
```
The long string value passed to `join_through` feels wrong here. Tried to find an example in the definitive documentation but haven't found one yet.

I would expect the syntax to be:

```elixir
many_to_many :users_following, Schema.User, join_through: "starred_messages", on_delete: :delete_all, on_replace: :delete
```

Since `on_delete` and `on_replace` should be keyword list options, not string values that need to be parsed an interpolated.

chrislaskey replied:
Also, it feels like there's a better name for this relationship. The term "Following" could mean a number of things. For instance, what if we introduce threaded replies to messages. A user who has replied once might be "following" that message.

What about a self-documenting name like `:starred_message_users`. It's a little verbose, but it would prevent any hesitation about the value.

chrislaskey commented on apps/chat_server/lib/chat_server/schemas/starred_message.ex:
```diff
@@ -0,0 +1,60 @@
+  def delete(%StarredMessage{} = starred_message) do
+    Repo.delete(starred_message)
+  end
+end
```
Looks great!

chrislaskey commented on apps/chat_server/priv/repo/migrations/20171003225928_add_starred_messages_table.exs:
```diff
@@ -0,0 +1,15 @@
+
+    create index(:starred_messages, [:message_id])
+    create index(:starred_messages, [:user_id])
+    create unique_index(:starred_messages, [:user_id, :message_id], name: :starred_messages_user_message_index)
```
Not a blocker, but one thing to be aware of with compound indexes like these is that often times the order is meaningful.

If a query is made asking for `user_id` then `message_id` then it'll use the index and do a fast lookup. If a query reverses the order and passes in `message_id` then `user_id` the index won't be used and a slower lookup will occur.

I'm not a database expert, so it's possible either Ecto optimizes to prevent an unoptimized query from happening, or that compound key ordering isn't an issue in PostgreSQL like it was in MySQL.

**Edit:** Just saw the next migration file adds a primary key. That sidesteps a lot of the compound index issue.

chrislaskey commented on apps/chat_server/priv/repo/migrations/20171025210428_add_id_to_starred_message_table.exs:
```diff
@@ -0,0 +1,9 @@
+      add :id, :binary_id, primary_key: true
+    end
+  end
+end
```
While building out a feature, the model often changes. It's up to you, but it's okay to go back and update the original migration file so there's just one instead of many. Not a blocker!

chrislaskey commented on apps/chat_server/test/chat_server/schemas/starred_message_test.exs:
```diff
@@ -0,0 +1,30 @@
+    test "Is valid when associated user and message are valid", %{user: user, message: message} do
+      params = %{user_id: user.id, message_id: message.id}
+      changeset = StarredMessage.changeset(%StarredMessage{}, params)
+       assert changeset.valid?
```
Looks like an extra space indentation here

chrislaskey commented on apps/chat_server/test/test_helper.exs:
```diff
@@ -1,5 +1,5 @@
 
 ExUnit.start()
+
+Ecto.Adapters.SQL.Sandbox.mode(ChatServer.Repo, :manual)
```
Thanks for adding a `pending` test pattern!

Could you also add back in the removed line:

```
{:ok, _} = Application.ensure_all_started(:ex_machina)
```

I think it got removed by accident when I merged in master. Appreciate it!

chrislaskey commented on apps/chat_websocket/lib/chat_websocket/channels/rooms_channel.ex:
```diff
@@ -20,11 +23,26 @@ defmodule ChatWebsocket.RoomsChannel do
   def handle_info(:after_join, socket) do
     %{user: user} = socket.assigns
 
+    push socket, "starred_messages_state", GetStarredMessages.call(user)
```
One pattern that changed since you wrote this is only using `_state` for Presence data. In other words, this should become:

```elixir
push socket, "starred_messages", GetStarredMessages.call(user)
```

chrislaskey commented:
Nice work! The feature is looking great. Most of my comments are minor syntax changes and Elixir language tweaks.

yourFriendWes committed:
fix merge conflict

yourFriendWes committed:
Updates based on PR feedback

- style changes and cleanup
- condense the two migrations for starred_messages into one

yourFriendWes committed:
WIP commit

- still not working when no starred_messages returned, hits poison
encoding error for empty array

yourFriendWes committed:
Revision to previous WIP commit

yourFriendWes committed:
Return success messages in star message handle_in functions

yourFriendWes committed:
Simplify GetStarredMessages.call

chrislaskey commented on apps/chat_server/lib/chat_server/contexts/starred_messages/create_starred_message.ex:
```diff
@@ -0,0 +1,28 @@
+
+  def call(message_id, user) do
+    with message <- Schema.Message.get(message_id),
+      {:ok, record} <- create_record(message, user) do
```
Looks great as a `with` statement. Minor note on formatting, consecutive `with` statements should be lined up with the first. In other words, this line should be indented 5 instead of 2:

```elixir
with message <- Schema.Message.get(message_id),
     {:ok, record} <- create_record(message, user) do
  {:ok, Schema.StarredMessage.preload(record, [:message, :user])}
```

chrislaskey committed:
Merge remote-tracking branch 'upstream/master' into add-starred-messages

chrislaskey committed:
Remove redundant `@derive`

yourFriendWes committed:
Merge pull request #3 from chrislaskey/add-starred-messages

Merge upstream/master

chrislaskey committed:
Fix minor indentation issue

chrislaskey reviewed (approved):
🌟 💯 🎉 💥 🎊
